pageserver/src/tenant/timeline.rs

```diff
@@ -491,6 +431,67 @@ impl Timeline {
         self.latest_gc_cutoff_lsn.read()
     }
 
+    pub fn get_context(&self, parent: &RequestContext) -> Result<RequestContext, TimelineState> {
+        let state_ref = self.state.borrow();
+        let state = *state_ref;
+
+        if state == TimelineState::Active || state == TimelineState::Suspended {
+            let mut cxt = RequestContext::with_parent(parent.task_kind, parent);
+            assert!(parent.timeline.is_none());
+            self.register_context(cxt.context_id, cxt.task_kind);
+            cxt.timeline = Some(self.myself.upgrade().unwrap());
+            Ok(cxt)
+        } else {
+            Err(state)
+        }
+    }
+
+    fn register_context(&self, context_id: u64, task_kind: TaskKind) {
+        if self
+            .active_contexts
+            .lock()
+            .unwrap()
+            .insert(context_id, task_kind)
+            .is_some()
+        {
+            panic!("active_contexts out of sync");
+        }
+    }
+
+    pub fn deregister_context(&self, context_id: u64) {
+        if self
+            .active_contexts
+            .lock()
+            .unwrap()
+            .remove(&context_id)
+            .is_none()
+        {
+            panic!("active_contexts out of sync");
+        }
+    }
+
+    ///
+    /// Wait until all RequestContexts registered with the Timeline have been dropped.
+    ///
+    /// This should be called only after setting the state to Stopping. Otherwise
+    /// new contexts can appear at any time.
+    ///
+    pub async fn wait_no_more_active_contexts(&self) {
```

Comment:
This method is not used anywhere.

Could all activeness checks be a part of the `context.rs::CONTEXTS` map (or `context.rs` module, at least?), since it's the same in `tenant.rs` too?